File: go/ontology/template_match.go
// SPDX-License-Identifier: Apache-2.0
package ontology

import (
	"context"
	"fmt"
	"sort"

	"github.com/jeffs-brain/memory/go/llm"
)

// Template matching constants ported from the intelligence service.
const (
	// TemplateMatchSemanticThreshold is the minimum cosine similarity for
	// a semantic embedding pair to count as a match during bipartite
	// greedy matching.
	TemplateMatchSemanticThreshold = 0.8

	// TemplateMatchCombinedMinimum is the minimum combined score
	// (overlap + coverage) / 2 required for a template to qualify as a
	// suggestion.
	TemplateMatchCombinedMinimum = 0.3
)

// TemplateSuggestion is the result of template matching.
type TemplateSuggestion struct {
	TemplateKey         string      `json:"templateKey"`
	TemplateLabel       string      `json:"templateLabel"`
	OverlapScore        float64     `json:"overlapScore"`
	CoverageScore       float64     `json:"coverageScore"`
	AdditionalTypes     []TypeEntry `json:"additionalTypes"`
	MissingFromTemplate []TypeEntry `json:"missingFromTemplate"`
}

// TemplateMatcherConfig configures the TemplateMatcher.
type TemplateMatcherConfig struct {
	// Embedder provides semantic embeddings. May be nil for exact-only matching.
	Embedder llm.Embedder

	// SemanticThreshold overrides TemplateMatchSemanticThreshold.
	// Zero uses the default (0.8).
	SemanticThreshold float64

	// CombinedMinimum overrides TemplateMatchCombinedMinimum.
	// Zero uses the default (0.3).
	CombinedMinimum float64
}

// TemplateMatcher suggests industry templates based on extracted types.
type TemplateMatcher struct {
	embedder          llm.Embedder
	semanticThreshold float64
	combinedMinimum   float64
	embeddingCache    map[string][][]float32
}

// NewTemplateMatcher creates a matcher. Embedder may be nil for exact-only matching.
func NewTemplateMatcher(embedder llm.Embedder) *TemplateMatcher {
	return NewTemplateMatcherWithConfig(TemplateMatcherConfig{Embedder: embedder})
}

// NewTemplateMatcherWithConfig creates a TemplateMatcher with configurable options.
func NewTemplateMatcherWithConfig(cfg TemplateMatcherConfig) *TemplateMatcher {
	semThresh := cfg.SemanticThreshold
	if semThresh == 0 {
		semThresh = TemplateMatchSemanticThreshold
	}
	combMin := cfg.CombinedMinimum
	if combMin == 0 {
		combMin = TemplateMatchCombinedMinimum
	}
	return &TemplateMatcher{
		embedder:          cfg.Embedder,
		semanticThreshold: semThresh,
		combinedMinimum:   combMin,
		embeddingCache:    make(map[string][][]float32),
	}
}

// Match finds the best-matching industry template for the given extraction.
// Uses semantic matching when an embedder is available, falling back to exact.
// Returns nil if no template scores above the combined threshold.
func (m *TemplateMatcher) Match(ctx context.Context, extracted ExtractionResult) (*TemplateSuggestion, error) {
	if len(extracted.NodeTypes) == 0 && len(extracted.EdgeTypes) == 0 {
		return nil, nil
	}

	if m.embedder == nil {
		return m.MatchExact(extracted), nil
	}

	return m.matchSemantic(ctx, extracted)
}

// MatchExact performs set-intersection-only matching (no embeddings).
// Returns nil if no template scores above the combined threshold.
func (m *TemplateMatcher) MatchExact(extracted ExtractionResult) *TemplateSuggestion {
	if len(extracted.NodeTypes) == 0 && len(extracted.EdgeTypes) == 0 {
		return nil
	}

	extractedNodeSet := make(map[string]TypeEntry, len(extracted.NodeTypes))
	for _, nt := range extracted.NodeTypes {
		extractedNodeSet[nt.Type] = nt
	}
	extractedEdgeSet := make(map[string]TypeEntry, len(extracted.EdgeTypes))
	for _, et := range extracted.EdgeTypes {
		extractedEdgeSet[et.Type] = et
	}

	extractedCount := len(extracted.NodeTypes) + len(extracted.EdgeTypes)

	var best *TemplateSuggestion
	var bestCombined float64

	keys := ListTemplates()
	for _, key := range keys {
		tmpl, ok := GetTemplate(key)
		if !ok {
			continue
		}

		templateNodeSet := make(map[string]TypeEntry, len(tmpl.NodeTypes))
		for _, nt := range tmpl.NodeTypes {
			templateNodeSet[nt.Type] = nt
		}
		templateEdgeSet := make(map[string]TypeEntry, len(tmpl.EdgeTypes))
		for _, et := range tmpl.EdgeTypes {
			templateEdgeSet[et.Type] = et
		}

		templateCount := len(tmpl.NodeTypes) + len(tmpl.EdgeTypes)
		if templateCount == 0 {
			continue
		}

		intersectionCount := 0
		for typeID := range extractedNodeSet {
			if _, ok := templateNodeSet[typeID]; ok {
				intersectionCount++
			}
		}
		for typeID := range extractedEdgeSet {
			if _, ok := templateEdgeSet[typeID]; ok {
				intersectionCount++
			}
		}

		overlapScore := 0.0
		if extractedCount > 0 {
			overlapScore = float64(intersectionCount) / float64(extractedCount)
		}
		coverageScore := float64(intersectionCount) / float64(templateCount)
		combined := (overlapScore + coverageScore) / 2

		if combined >= m.combinedMinimum && combined > bestCombined {
			additional := computeAdditionalTypes(extracted, templateNodeSet, templateEdgeSet)
			missing := computeMissingTypes(tmpl, extractedNodeSet, extractedEdgeSet)

			bestCombined = combined
			best = &TemplateSuggestion{
				TemplateKey:         key,
				TemplateLabel:       tmpl.Label,
				OverlapScore:        overlapScore,
				CoverageScore:       coverageScore,
				AdditionalTypes:     additional,
				MissingFromTemplate: missing,
			}
		}
	}

	return best
}

func (m *TemplateMatcher) matchSemantic(ctx context.Context, extracted ExtractionResult) (*TemplateSuggestion, error) {
	extractedTypes := combineTypes(extracted.NodeTypes, extracted.EdgeTypes)
	if len(extractedTypes) == 0 {
		return nil, nil
	}

	extractedTexts := make([]string, len(extractedTypes))
	for i, t := range extractedTypes {
		extractedTexts[i] = t.Label + ": " + t.Description
	}

	extractedEmbeddings, err := m.embedder.Embed(ctx, extractedTexts)
	if err != nil {
		return nil, fmt.Errorf("ontology: embed extracted types: %w", err)
	}

	var best *TemplateSuggestion
	var bestCombined float64

	keys := ListTemplates()
	for _, key := range keys {
		tmpl, ok := GetTemplate(key)
		if !ok {
			continue
		}

		templateTypes := combineTypes(tmpl.NodeTypes, tmpl.EdgeTypes)
		if len(templateTypes) == 0 {
			continue
		}

		var templateEmbeddings [][]float32
		if cached, ok := m.embeddingCache[key]; ok {
			templateEmbeddings = cached
		} else {
			templateTexts := make([]string, len(templateTypes))
			for i, t := range templateTypes {
				templateTexts[i] = t.Label + ": " + t.Description
			}

			var embedErr error
			templateEmbeddings, embedErr = m.embedder.Embed(ctx, templateTexts)
			if embedErr != nil {
				return nil, fmt.Errorf("ontology: embed template %q types: %w", key, embedErr)
			}
			m.embeddingCache[key] = templateEmbeddings
		}

		matchCount := greedyBipartiteMatch(extractedEmbeddings, templateEmbeddings, m.semanticThreshold)

		extractedCount := len(extractedTypes)
		templateCount := len(templateTypes)

		overlapScore := 0.0
		if extractedCount > 0 {
			overlapScore = float64(matchCount) / float64(extractedCount)
		}
		coverageScore := 0.0
		if templateCount > 0 {
			coverageScore = float64(matchCount) / float64(templateCount)
		}
		combined := (overlapScore + coverageScore) / 2

		if combined >= m.combinedMinimum && combined > bestCombined {
			extractedNodeSet := make(map[string]TypeEntry, len(extracted.NodeTypes))
			for _, nt := range extracted.NodeTypes {
				extractedNodeSet[nt.Type] = nt
			}
			extractedEdgeSet := make(map[string]TypeEntry, len(extracted.EdgeTypes))
			for _, et := range extracted.EdgeTypes {
				extractedEdgeSet[et.Type] = et
			}
			templateNodeSet := make(map[string]TypeEntry, len(tmpl.NodeTypes))
			for _, nt := range tmpl.NodeTypes {
				templateNodeSet[nt.Type] = nt
			}
			templateEdgeSet := make(map[string]TypeEntry, len(tmpl.EdgeTypes))
			for _, et := range tmpl.EdgeTypes {
				templateEdgeSet[et.Type] = et
			}

			additional := computeAdditionalTypes(extracted, templateNodeSet, templateEdgeSet)
			missing := computeMissingTypes(tmpl, extractedNodeSet, extractedEdgeSet)

			bestCombined = combined
			best = &TemplateSuggestion{
				TemplateKey:         key,
				TemplateLabel:       tmpl.Label,
				OverlapScore:        overlapScore,
				CoverageScore:       coverageScore,
				AdditionalTypes:     additional,
				MissingFromTemplate: missing,
			}
		}
	}

	return best, nil
}

// greedyBipartiteMatch counts the number of matched pairs between
// extracted and template embeddings using greedy best-first matching.
// Each extracted embedding is matched to the most-similar template
// embedding with cosine similarity >= threshold. Once matched, a
// template embedding is removed from the candidate pool.
func greedyBipartiteMatch(extractedVecs, templateVecs [][]float32, threshold float64) int {
	if len(extractedVecs) == 0 || len(templateVecs) == 0 {
		return 0
	}

	type scoredPair struct {
		extractedIdx int
		templateIdx  int
		similarity   float64
	}

	var pairs []scoredPair
	for i, ev := range extractedVecs {
		for j, tv := range templateVecs {
			sim, err := CosineSimilarity(ev, tv)
			if err != nil || sim < threshold {
				continue
			}
			pairs = append(pairs, scoredPair{i, j, sim})
		}
	}

	sort.Slice(pairs, func(a, b int) bool {
		return pairs[a].similarity > pairs[b].similarity
	})

	usedExtracted := make(map[int]bool)
	usedTemplate := make(map[int]bool)
	matchCount := 0

	for _, p := range pairs {
		if usedExtracted[p.extractedIdx] || usedTemplate[p.templateIdx] {
			continue
		}
		usedExtracted[p.extractedIdx] = true
		usedTemplate[p.templateIdx] = true
		matchCount++
	}

	return matchCount
}

// computeAdditionalTypes returns types in extracted that are not in the template.
func computeAdditionalTypes(extracted ExtractionResult, templateNodeSet, templateEdgeSet map[string]TypeEntry) []TypeEntry {
	var additional []TypeEntry
	for _, nt := range extracted.NodeTypes {
		if _, ok := templateNodeSet[nt.Type]; !ok {
			additional = append(additional, nt)
		}
	}
	for _, et := range extracted.EdgeTypes {
		if _, ok := templateEdgeSet[et.Type]; !ok {
			additional = append(additional, et)
		}
	}
	if additional == nil {
		return []TypeEntry{}
	}
	return additional
}

// computeMissingTypes returns types in the template that are not in extracted.
func computeMissingTypes(tmpl IndustryTemplate, extractedNodeSet, extractedEdgeSet map[string]TypeEntry) []TypeEntry {
	var missing []TypeEntry
	for _, nt := range tmpl.NodeTypes {
		if _, ok := extractedNodeSet[nt.Type]; !ok {
			missing = append(missing, nt)
		}
	}
	for _, et := range tmpl.EdgeTypes {
		if _, ok := extractedEdgeSet[et.Type]; !ok {
			missing = append(missing, et)
		}
	}
	if missing == nil {
		return []TypeEntry{}
	}
	return missing
}

// combineTypes merges node and edge type entries into a single slice.
func combineTypes(nodeTypes, edgeTypes []TypeEntry) []TypeEntry {
	combined := make([]TypeEntry, 0, len(nodeTypes)+len(edgeTypes))
	combined = append(combined, nodeTypes...)
	combined = append(combined, edgeTypes...)
	return combined
}

File: sdks/ts/memory/src/ontology/template-match.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Template matching: score extracted ontology types against all 6
 * industry templates and suggest the best match. Supports exact set
 * intersection and semantic embedding similarity with greedy bipartite
 * matching.
 *
 * Port of apps/intelligence-service/src/services/document-processing/
 * ontology-extraction.service.ts (matchAgainstTemplatesExact and
 * matchAgainstTemplatesSemantic).
 */

import type { Embedder } from '../llm/types.js'
import type { TypeEntry, IndustryTemplate } from './templates.js'
import type { ExtractionResult } from './extract.js'

import { listTemplates, getTemplate } from './templates.js'
import { cosineSimilarity } from './similarity.js'

export type { ExtractionResult } from './extract.js'

/** Minimum cosine similarity for a semantic embedding pair to count as a match. */
export const TEMPLATE_MATCH_SEMANTIC_THRESHOLD = 0.8

/** Minimum combined score (overlap + coverage) / 2 for a suggestion. */
export const TEMPLATE_MATCH_COMBINED_MINIMUM = 0.3

export type TemplateSuggestion = {
  readonly templateKey: string
  readonly templateLabel: string
  readonly overlapScore: number
  readonly coverageScore: number
  readonly additionalTypes: TypeEntry[]
  readonly missingFromTemplate: TypeEntry[]
}

export type TemplateMatcherOptions = {
  readonly embedder?: Embedder
  /** Override the default semantic threshold (0.8). */
  readonly semanticThreshold?: number
  /** Override the default combined minimum (0.3). */
  readonly combinedMinimum?: number
}

/**
 * TemplateMatcher suggests industry templates based on extracted types.
 * Uses exact set intersection by default, and semantic embedding similarity
 * when an embedder is provided.
 */
export class TemplateMatcher {
  private readonly embedder: Embedder | undefined
  private readonly semanticThreshold: number
  private readonly combinedMinimum: number
  private readonly embeddingCache = new Map<string, readonly (readonly number[])[]>()

  constructor(options: TemplateMatcherOptions) {
    this.embedder = options.embedder
    this.semanticThreshold = options.semanticThreshold ?? TEMPLATE_MATCH_SEMANTIC_THRESHOLD
    this.combinedMinimum = options.combinedMinimum ?? TEMPLATE_MATCH_COMBINED_MINIMUM
  }

  /**
   * Find the best-matching industry template for the given extraction.
   * Uses semantic matching when an embedder is available, falling back to exact.
   * Returns undefined if no template scores above the combined threshold.
   */
  async match(
    extracted: ExtractionResult,
    signal?: AbortSignal,
  ): Promise<TemplateSuggestion | undefined> {
    if (extracted.nodeTypes.length === 0 && extracted.edgeTypes.length === 0) {
      return undefined
    }

    if (this.embedder === undefined) {
      return this.matchExact(extracted)
    }

    return this.matchSemantic(extracted, signal)
  }

  /**
   * Perform set-intersection-only matching (no embeddings).
   * Returns undefined if no template scores above the combined threshold.
   */
  matchExact(extracted: ExtractionResult): TemplateSuggestion | undefined {
    if (extracted.nodeTypes.length === 0 && extracted.edgeTypes.length === 0) {
      return undefined
    }

    const extractedNodeSet = new Map<string, TypeEntry>()
    for (const nt of extracted.nodeTypes) {
      extractedNodeSet.set(nt.type, nt)
    }
    const extractedEdgeSet = new Map<string, TypeEntry>()
    for (const et of extracted.edgeTypes) {
      extractedEdgeSet.set(et.type, et)
    }

    const extractedCount = extracted.nodeTypes.length + extracted.edgeTypes.length

    let best: TemplateSuggestion | undefined
    let bestCombined = 0

    const keys = listTemplates()
    for (const key of keys) {
      const tmpl = getTemplate(key)
      if (tmpl === undefined) continue

      const templateNodeSet = new Map<string, TypeEntry>()
      for (const nt of tmpl.nodeTypes) {
        templateNodeSet.set(nt.type, nt)
      }
      const templateEdgeSet = new Map<string, TypeEntry>()
      for (const et of tmpl.edgeTypes) {
        templateEdgeSet.set(et.type, et)
      }

      const templateCount = tmpl.nodeTypes.length + tmpl.edgeTypes.length
      if (templateCount === 0) continue

      let intersectionCount = 0
      for (const typeId of extractedNodeSet.keys()) {
        if (templateNodeSet.has(typeId)) intersectionCount++
      }
      for (const typeId of extractedEdgeSet.keys()) {
        if (templateEdgeSet.has(typeId)) intersectionCount++
      }

      const overlapScore = extractedCount > 0 ? intersectionCount / extractedCount : 0
      const coverageScore = intersectionCount / templateCount
      const combined = (overlapScore + coverageScore) / 2

      if (combined >= this.combinedMinimum && combined > bestCombined) {
        const additional = computeAdditionalTypes(extracted, templateNodeSet, templateEdgeSet)
        const missing = computeMissingTypes(tmpl, extractedNodeSet, extractedEdgeSet)

        bestCombined = combined
        best = {
          templateKey: key,
          templateLabel: tmpl.label,
          overlapScore,
          coverageScore,
          additionalTypes: additional,
          missingFromTemplate: missing,
        }
      }
    }

    return best
  }

  private async matchSemantic(
    extracted: ExtractionResult,
    signal?: AbortSignal,
  ): Promise<TemplateSuggestion | undefined> {
    const extractedTypes = combineTypes(extracted.nodeTypes, extracted.edgeTypes)
    if (extractedTypes.length === 0) return undefined

    const extractedTexts = extractedTypes.map((t) => `${t.label}: ${t.description}`)
    const extractedEmbeddings = await this.embedder!.embed(extractedTexts, signal)

    let best: TemplateSuggestion | undefined
    let bestCombined = 0

    const keys = listTemplates()
    for (const key of keys) {
      const tmpl = getTemplate(key)
      if (tmpl === undefined) continue

      const templateTypes = combineTypes(tmpl.nodeTypes, tmpl.edgeTypes)
      if (templateTypes.length === 0) continue

      let templateEmbeddings = this.embeddingCache.get(key)
      if (templateEmbeddings === undefined) {
        const templateTexts = templateTypes.map((t) => `${t.label}: ${t.description}`)
        templateEmbeddings = await this.embedder!.embed(templateTexts, signal)
        this.embeddingCache.set(key, templateEmbeddings)
      }

      const matchCount = greedyBipartiteMatch(
        extractedEmbeddings,
        templateEmbeddings,
        this.semanticThreshold,
      )

      const extractedCount = extractedTypes.length
      const templateCount = templateTypes.length

      const overlapScore = extractedCount > 0 ? matchCount / extractedCount : 0
      const coverageScore = templateCount > 0 ? matchCount / templateCount : 0
      const combined = (overlapScore + coverageScore) / 2

      if (combined >= this.combinedMinimum && combined > bestCombined) {
        const extractedNodeSet = new Map<string, TypeEntry>()
        for (const nt of extracted.nodeTypes) {
          extractedNodeSet.set(nt.type, nt)
        }
        const extractedEdgeSet = new Map<string, TypeEntry>()
        for (const et of extracted.edgeTypes) {
          extractedEdgeSet.set(et.type, et)
        }
        const templateNodeSet = new Map<string, TypeEntry>()
        for (const nt of tmpl.nodeTypes) {
          templateNodeSet.set(nt.type, nt)
        }
        const templateEdgeSet = new Map<string, TypeEntry>()
        for (const et of tmpl.edgeTypes) {
          templateEdgeSet.set(et.type, et)
        }

        const additional = computeAdditionalTypes(extracted, templateNodeSet, templateEdgeSet)
        const missing = computeMissingTypes(tmpl, extractedNodeSet, extractedEdgeSet)

        bestCombined = combined
        best = {
          templateKey: key,
          templateLabel: tmpl.label,
          overlapScore,
          coverageScore,
          additionalTypes: additional,
          missingFromTemplate: missing,
        }
      }
    }

    return best
  }
}

/**
 * Greedy bipartite matching: for each pair, compute cosine similarity.
 * Sort by similarity descending, greedily assign each pair if both sides
 * are unmatched. Only counts pairs with similarity >= threshold.
 */
function greedyBipartiteMatch(
  extractedVecs: readonly (readonly number[])[],
  templateVecs: readonly (readonly number[])[],
  threshold: number,
): number {
  if (extractedVecs.length === 0 || templateVecs.length === 0) return 0

  const pairs: Array<{ extractedIdx: number; templateIdx: number; similarity: number }> = []

  for (let i = 0; i < extractedVecs.length; i++) {
    const ev = extractedVecs[i]
    if (ev === undefined) continue
    for (let j = 0; j < templateVecs.length; j++) {
      const tv = templateVecs[j]
      if (tv === undefined) continue
      try {
        const sim = cosineSimilarity(ev, tv)
        if (sim >= threshold) {
          pairs.push({ extractedIdx: i, templateIdx: j, similarity: sim })
        }
      } catch {
        // Skip if vectors have different lengths
      }
    }
  }

  pairs.sort((a, b) => b.similarity - a.similarity)

  const usedExtracted = new Set<number>()
  const usedTemplate = new Set<number>()
  let matchCount = 0

  for (const p of pairs) {
    if (usedExtracted.has(p.extractedIdx) || usedTemplate.has(p.templateIdx)) continue
    usedExtracted.add(p.extractedIdx)
    usedTemplate.add(p.templateIdx)
    matchCount++
  }

  return matchCount
}

function computeAdditionalTypes(
  extracted: ExtractionResult,
  templateNodeSet: Map<string, TypeEntry>,
  templateEdgeSet: Map<string, TypeEntry>,
): TypeEntry[] {
  const additional: TypeEntry[] = []
  for (const nt of extracted.nodeTypes) {
    if (!templateNodeSet.has(nt.type)) additional.push(nt)
  }
  for (const et of extracted.edgeTypes) {
    if (!templateEdgeSet.has(et.type)) additional.push(et)
  }
  return additional
}

function computeMissingTypes(
  tmpl: IndustryTemplate,
  extractedNodeSet: Map<string, TypeEntry>,
  extractedEdgeSet: Map<string, TypeEntry>,
): TypeEntry[] {
  const missing: TypeEntry[] = []
  for (const nt of tmpl.nodeTypes) {
    if (!extractedNodeSet.has(nt.type)) missing.push(nt)
  }
  for (const et of tmpl.edgeTypes) {
    if (!extractedEdgeSet.has(et.type)) missing.push(et)
  }
  return missing
}

function combineTypes(
  nodeTypes: readonly TypeEntry[],
  edgeTypes: readonly TypeEntry[],
): TypeEntry[] {
  return [...nodeTypes, ...edgeTypes]
}
